File: leetcode/MinimumValueToGetPositiveStepByStepSum.py
from typing import List


class Solution:
    def minStartValue(self, nums: List[int]) -> int:
        """
        Calculates the minimum positive startValue such that the step-by-step sum
        is never less than 1.

        This is solved by finding the minimum prefix sum during a single pass.
        The startValue must be large enough to ensure this minimum sum is at least 1.
        """
        # We need to find the minimum possible value of the running sum.
        # Let's track the running sum and its minimum value encountered so far.
        min_running_sum = 0  # Initialize to 0, handles cases where all sums are positive.
        running_sum = 0

        for num in nums:
            running_sum += num
            min_running_sum = min(min_running_sum, running_sum)

        # Let the minimum start value be `s`.
        # We need the condition `s + running_sum >= 1` to hold for all steps.
        # This is equivalent to `s + min_running_sum >= 1`.
        # Therefore, the minimum `s` must be `1 - min_running_sum`.
        #
        # If `min_running_sum` is 0 (or positive), the result is 1.
        # If `min_running_sum` is negative (e.g., -5), the result is `1 - (-5) = 6`.
        return 1 - min_running_sum

File: leetcode/easy/MinimumValueToGetPositiveStepByStepSum.java
/**
 * Solution for LeetCode problem "Minimum Value to Get Positive Step by Step Sum".
 * This class provides a method to find the minimum positive `startValue` such that
 * the step-by-step sum of `startValue` plus elements in `nums` is never less than 1.
 */
class Solution {

    /**
     * Calculates the minimum positive `startValue` such that the step-by-step sum
     * of `startValue` plus elements in `nums` is always greater than or equal to 1.
     *
     * The approach is to track the minimum running sum encountered while iterating through `nums`.
     * Let `running_sum` be the sum of `startValue` and the elements of `nums` up to the current point.
     * We want `running_sum >= 1` at all times.
     * This is equivalent to `startValue + (sum of nums elements up to current point) >= 1`.
     * So, `startValue >= 1 - (sum of nums elements up to current point)`.
     *
     * To satisfy this for all steps, `startValue` must be greater than or equal to
     * `1 - (minimum sum of nums elements encountered during traversal)`.
     * If the minimum sum of `nums` elements is `min_prefix_sum` (which can be negative),
     * then `startValue` must be at least `1 - min_prefix_sum`.
     *
     * @param nums An array of integers.
     * @return The minimum positive `startValue`.
     */
    public int minStartValue(int[] nums) {
        int minRunningSum = 0;
        int runningSum = 0;

        for (int num : nums) {
            runningSum += num;
            minRunningSum = Math.min(minRunningSum, runningSum);
        }

        return 1 - minRunningSum;
    }
}
